File: src/pipeline/stages.rs
/// Stages for workspace creation pipeline.
#[derive(Debug, Clone, Copy, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
pub enum CreateStage {
    Validate,
    Provision,
    Infra,
    Source,
    Configure,
    Setup,
    Network,
}

impl CreateStage {
    pub fn all() -> &'static [CreateStage] {
        &[
            Self::Validate,
            Self::Provision,
            Self::Infra,
            Self::Source,
            Self::Configure,
            Self::Setup,
            Self::Network,
        ]
    }

    pub fn label(&self) -> &'static str {
        match self {
            Self::Validate => "Validating config and hosts",
            Self::Provision => "Allocating IP and slots",
            Self::Infra => "Starting shared services and databases",
            Self::Source => "Creating git worktrees",
            Self::Configure => "Generating compose overrides and env files",
            Self::Setup => "Running setup commands",
            Self::Network => "Creating docker network",
        }
    }
}

/// Stages for workspace deletion pipeline.
#[derive(Debug, Clone, Copy, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
pub enum DeleteStage {
    Stop,
    Release,
    Cleanup,
    Remove,
    Finalize,
}

impl DeleteStage {
    pub fn all() -> &'static [DeleteStage] {
        &[
            Self::Stop,
            Self::Release,
            Self::Cleanup,
            Self::Remove,
            Self::Finalize,
        ]
    }

    pub fn label(&self) -> &'static str {
        match self {
            Self::Stop => "Stopping services",
            Self::Release => "Releasing IP and slots",
            Self::Cleanup => "Running pre-delete commands",
            Self::Remove => "Removing worktrees and databases",
            Self::Finalize => "Cleaning up network and folders",
        }
    }
}

File: internal/pipeline/stages.go
package pipeline

type CreateStage int

const (
	StageValidate CreateStage = iota
	StageProvision
	StageInfra
	StageSource
	StageConfigure
	StageSetup
	StageNetwork
)

var AllCreateStages = []CreateStage{
	StageValidate, StageProvision, StageInfra, StageSource,
	StageConfigure, StageSetup, StageNetwork,
}

func (s CreateStage) Label() string {
	switch s {
	case StageValidate:
		return "Validating config and hosts"
	case StageProvision:
		return "Allocating IP and slots"
	case StageInfra:
		return "Starting shared services and databases"
	case StageSource:
		return "Creating git worktrees (parallel)"
	case StageConfigure:
		return "Configuring repos (parallel)"
	case StageSetup:
		return "Running setup commands (parallel)"
	case StageNetwork:
		return "Creating docker network"
	default:
		return "Unknown"
	}
}

type DeleteStage int

const (
	StageStop DeleteStage = iota
	StageRelease
	StageCleanup
	StageRemove
	StageFinalize
)

var AllDeleteStages = []DeleteStage{
	StageStop, StageRelease, StageCleanup, StageRemove, StageFinalize,
}

func (s DeleteStage) Label() string {
	switch s {
	case StageStop:
		return "Stopping services"
	case StageRelease:
		return "Releasing IP and slots"
	case StageCleanup:
		return "Running pre-delete commands"
	case StageRemove:
		return "Removing worktrees and databases"
	case StageFinalize:
		return "Cleaning up network and folders"
	default:
		return "Unknown"
	}
}
